solidity/contracts/BondedECDSAKeepVendor.sol

```diff
@@ -140,123 +154,31 @@ contract BondedECDSAKeepVendor is Proxy {
-    /// EIP-1967.
-    function assertSlot(bytes32 slot, bytes memory key) internal pure {
-        assert(slot == bytes32(uint256(keccak256(key)) - 1));
-    }
-
-    /// @notice Gets the address of the current vendor implementation.
+    /// @notice Gets the address of the current vendor implementation address.
     /// @return Address of the current implementation.
     function implementation() public view returns (address) {
         return _implementation();
     }
 
-    /// @dev Returns the current implementation. Implements function from `Proxy`
-    /// contract.
+    /// @dev Returns the current implementation address. Implements function
+    /// from `Proxy` contract.
     /// @return Address of the current implementation
-    function _implementation() internal view returns (address impl) {
-        bytes32 slot = IMPLEMENTATION_SLOT;
-        /* solium-disable-next-line */
-        assembly {
-            impl := sload(slot)
-        }
-    }
-
-    /// @notice Sets the address of the current implementation.
-    /// @param _implementation Address representing the new implementation to
-    /// be set.
-    function setImplementation(address _implementation) internal {
-        bytes32 slot = IMPLEMENTATION_SLOT;
-        /* solium-disable-next-line */
-        assembly {
-            sstore(slot, _implementation)
-        }
-    }
-
-    function upgradeTimeDelay()
-        public
-        view
-        returns (uint256 _upgradeTimeDelay)
-    {
-        bytes32 position = UPGRADE_TIME_DELAY_SLOT;
-        /* solium-disable-next-line */
-        assembly {
-            _upgradeTimeDelay := sload(position)
-        }
-    }
-
-    function setUpgradeTimeDelay(uint256 _upgradeTimeDelay) internal {
-        bytes32 position = UPGRADE_TIME_DELAY_SLOT;
-        /* solium-disable-next-line */
-        assembly {
-            sstore(position, _upgradeTimeDelay)
-        }
-    }
-
-    function newImplementation()
-        public
-        view
-        returns (address _newImplementation)
-    {
-        bytes32 position = UPGRADE_IMPLEMENTATION_SLOT;
-        /* solium-disable-next-line */
-        assembly {
-            _newImplementation := sload(position)
-        }
-    }
-
-    function setNewImplementation(address _newImplementation) internal {
-        bytes32 position = UPGRADE_IMPLEMENTATION_SLOT;
-        /* solium-disable-next-line */
-        assembly {
-            sstore(position, _newImplementation)
-        }
-    }
-
-    function upgradeInitiatedTimestamp()
-        public
-        view
-        returns (uint256 _upgradeInitiatedTimestamp)
-    {
-        bytes32 position = UPGRADE_INIT_TIMESTAMP_SLOT;
-        /* solium-disable-next-line */
-        assembly {
-            _upgradeInitiatedTimestamp := sload(position)
-        }
-    }
+    function _implementation() internal view returns (address) {
+        Implementation memory implementation = getImplementation(
+            currentImplementationID()
+        );
```

Comment:
Do we know what's the impact on the gas cost? It affects all the calls to the system. In the previous implementation we were doing a single `sload`, without reading the entire map key and calling other functions.

I think the most effective and simplest solution is to store the current implementation data as before with just
```
bytes32 slot = IMPLEMENTATION_SLOT;
/* solium-disable-next-line */
assembly {
  impl := sload(slot)
}
```
here and have a `mapping` between the new version ID and initialization call data. When the upgrade starts, we commit to initialization and add the data to the mapping. Then, when the upgrade completes we execute the initialization using the data we previously committed to. If one wants to go through the history of upgrades they can use events.